loadgen/generic_executor.go

```diff
@@ -0,0 +1,136 @@
+package loadgen
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.temporal.io/sdk/client"
+	"go.uber.org/zap"
+)
+
+type GenericExecutor struct {
+	// Function to execute a single iteration of this scenario
+	Execute func(ctx context.Context, run *Run) error
+	// Default configuration if any.
+	DefaultConfiguration RunConfiguration
+}
+
+func (g GenericExecutor) GetDefaultConfiguration() RunConfiguration { return g.DefaultConfiguration }
+
+type genericRun struct {
+	executor *GenericExecutor
+	options  RunOptions
+	config   RunConfiguration
+	logger   *zap.SugaredLogger
+	// Timer capturing E2E execution of each scenario run iteration.
+	executeTimer client.MetricsTimer
+}
+
+// Run a scenario.
+func (g GenericExecutor) Run(ctx context.Context, options RunOptions) error {
+	r, err := g.newRun(options)
+	if err != nil {
+		return err
+	}
+	return r.Run(ctx)
+}
+
+func (g *GenericExecutor) newRun(options RunOptions) (*genericRun, error) {
+	run := &genericRun{
+		executor: g,
+		options:  options,
+		config:   options.Configuration,
+		logger:   options.Logger,
+		executeTimer: options.MetricsHandler.WithTags(
+			map[string]string{"scenario": options.ScenarioName}).Timer("omes_execute_histogram"),
+	}
+
+	// Setup config
+	if run.config.Duration == 0 && run.config.Iterations == 0 {
+		run.config.Duration, run.config.Iterations = g.DefaultConfiguration.Duration, g.DefaultConfiguration.Iterations
+	}
+	if run.config.MaxConcurrent == 0 {
+		run.config.MaxConcurrent = g.DefaultConfiguration.MaxConcurrent
+	}
+	run.config.ApplyDefaults()
+	if run.config.Iterations > 0 && run.config.Duration > 0 {
+		return nil, fmt.Errorf("invalid scenario: iterations and duration are mutually exclusive")
+	}
+
+	return run, nil
+}
+
+// Run a scenario.
+// Spins up coroutines according to the scenario configuration.
+// Each coroutine runs the scenario Execute method in a loop until the scenario duration or max iterations is reached.
+func (g *genericRun) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	if g.config.Duration > 0 {
+		ctx, cancel = context.WithTimeout(ctx, g.config.Duration)
+	}
+	defer cancel()
+
+	startTime := time.Now()
+	var runErr error
+	doneCh := make(chan error)
+	var currentlyRunning int
```

Comment:
nit: I'd name this `numCurrentlyRunning` to clarify this is a count not a toggle state.

Reply:
Not what a Go programmer would usually say, heh. You're lucky it's not `c` :-)

Reply:
Definitely lucky it's not C